standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -5390,6 +5406,493 @@ public void countOpenTxns() throws MetaException {
+                                                 IHMSHandler handler,
+                                                 List<MetaStoreEventListener> listeners,
+                                                 Table tbl,
+                                                 String validWriteIds, long writeId) throws MetaException {
+    Connection dbConn = null;
+    String catName = tbl.getCatName();
+    String dbName = tbl.getDbName();
+    String tableName = tbl.getTableName();
+    boolean isAcidTable = TxnUtils.isAcidTable(tbl);
+
+    LOG.info("ETL_PERF started updatePartitionColumnStatistics");
+    boolean committed = false;
+    try {
+      try {
+        lockInternal();
+        dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED, connPoolMutex);
+
+        LOG.info("ETL_PERF started getPartitionInfo");
+        List<String> partNames = partColStatsMap.keySet().stream().map(
+                e -> quoteString(e)).collect(Collectors.toList()
+        );
+        Map<String, PartitionInfo> partitionInfoMap = getPartitionInfo(dbConn, catName, dbName, tableName, partNames);
+        LOG.info("ETL_PERF done getPartitionInfo");
+
+        List<Long> partIdList = partitionInfoMap.values().stream().map(
+                e -> e.partitionId).collect(Collectors.toList()
+        );
+
+        LOG.info("ETL_PERF started updatePartitionParamTable");
+        Map<String, Map<String, String>> result =
+                updatePartitionParamTable(dbConn, partitionInfoMap, partColStatsMap,
+                        partIdList, validWriteIds, writeId, isAcidTable);
+        LOG.info("ETL_PERF done updatePartitionParamTable");
+
+        LOG.info("ETL_PERF started cleanOldStatsFromPartColStatTable ");
+        cleanOldStatsFromPartColStatTable(partitionInfoMap, partColStatsMap, dbConn);
+        LOG.info("ETL_PERF done cleanOldStatsFromPartColStatTable ");
+
+        LOG.info("ETL_PERF started insertIntoPartColStatTable ");
+        insertIntoPartColStatTable(partitionInfoMap, partColStatsMap, dbConn);
+        LOG.info("ETL_PERF done insertIntoPartColStatTable ");
+
+        LOG.info("ETL_PERF started notifyEventWithDirectSql");
+        for (Map.Entry entry : result.entrySet()) {
+          Map<String, String> parameters = (Map<String, String>) entry.getValue();
+          ColumnStatistics colStats = partColStatsMap.get(entry.getKey());
+          List<String> partVals = getPartValsFromName(tbl, colStats.getStatsDesc().getPartName());
+          if (transactionalListeners != null) {
+            MetaStoreListenerNotifier.notifyEventWithDirectSql(transactionalListeners,
+                    EventMessage.EventType.UPDATE_PARTITION_COLUMN_STAT,
+                    new UpdatePartitionColumnStatEvent(colStats, partVals, parameters,
+                            tbl, writeId, handler), dbConn, sqlGenerator);
+          }
+
+          if (listeners != null) {
+            MetaStoreListenerNotifier.notifyEvent(listeners,
```

Comment:
should we pass batch of events to the dbNotificationListener as well?

Reply:
That will be done as a separate Jira as this require some more thinking.